### ISSUE TITLE
Incorrect results for nested cardinality aggregations

### ISSUE THREAD

**Issue by chrisdrusso** [elastic/elasticsearch#18499]
**Elasticsearch version**: 
2.3.0

**JVM version**: 
1.8.0_40
1.8.0_91

**OS version**: 
Windows 10

**Description of the problem including expected versus actual behavior**:
When nesting aggregations of term cardinality (unique counts of terms in Kibana), the results are often incorrect.  Depending on the data, sometimes the result is higher than it should be, other times it is lower than it should be.  When the nesting is removed, the results for a single aggregation of term cardinality is correct.  

In the following example, the results show 8 distinct values for field3, but there are only 7 unique values for field3.  

**Steps to reproduce**:
 1. Index the following data:
`POST test/_bulk
{"index":{"_index":"test","_type":"test"}}
{"field1":"A","field2":"B","field3":"1"}
{"index":{"_index":"test","_type":"test"}}
{"field1":"A","field2":"B","field3":"2"}
{"index":{"_index":"test","_type":"test"}}
{"field1":"A","field2":"B","field3":"3"}
{"index":{"_index":"test","_type":"test"}}
{"field1":"A","field2":"B","field3":"4"}
{"index":{"_index":"test","_type":"test"}}
{"field1":"A","field2":"B","field3":"5"}
{"index":{"_index":"test","_type":"test"}}
{"field1":"A","field2":"B","field3":"6"}
{"index":{"_index":"test","_type":"test"}}
{"field1":"A","field2":"B","field3":"7"}`
 2. Run the following aggregation query:
`POST test/_search
{
  "query": {
    "filtered": {
      "query": {
        "query_string": {
          "analyze_wildcard": true,
          "query": "*"
        }
      },
      "filter": {
        "bool": {
          "must": [
            {
              "query": {
                "query_string": {
                  "analyze_wildcard": true,
                  "query": "*"
                }
              }
            }
          ],
          "must_not": []
        }
      }
    }
  },
  "size": 0,
  "aggs": {
    "3": {
      "terms": {
        "field": "field1",
        "size": 0,
        "order": {
          "_term": "asc"
        }
      },
      "aggs": {
        "4": {
          "terms": {
            "field": "field2",
            "size": 0,
            "order": {
              "_term": "asc"
            }
          },
          "aggs": {
            "1": {
              "cardinality": {
                "field": "field3"
              }
            }
          }
        }
      }
    }
  }
}`

**Provide logs (if relevant)**:
Nothing in the logs.

**Comment by jimferenczi** [comment]
This is because the precision threshold depends on the number of parent aggregations. We expect to have less values in the child aggregations so the threshold is lower:
https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-metrics-cardinality-aggregation.html
Of course you can control it:
https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-metrics-cardinality-aggregation.html#_precision_control

Bottom line is that the counts are approximate to save memory and that you can control how approximate they are by setting precision_threshold.

**Comment by chrisdrusso** [comment]
Worked perfectly.  Thank you jimferenczi.
